[elemental_tv] Add elemental_tv extractor

mp107 opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### This pull requests adds an extractor for a bulgarian website https://elemental.tv

Please let me know if there are improvements to be made to the code before merging!

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

glensc commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        # Use double quotes (") as server returns error 400 while using apostrophe (')
+        post_data = str(post_data).replace("'", '"').encode(encoding='UTF-8')
```
you have an apostrophe in your username or password where it fails?

seems you use `str()` to encode dict to json. of course it fails, as that's not valid json. json standard has double quotes. 

use proper function to encode json:

```py
import json
post_data = json.dumps(post_data)
```

glensc commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,117 @@
+        # Use double quotes (") as server returns error 400 while using apostrophe (')
+        post_data = str(post_data).replace("'", '"').encode(encoding='UTF-8')
+
+        res_api = self._download_json(url, self.channel_id, data=post_data)
```
```suggestion
        res_api = self._download_json(url, self.channel_id, data=post_data).get('data')
```

repeated `.get('data')` calls. perhaps store them in res_api instead?

glensc commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,117 @@
+        result = {
+            'id': self.channel_id,
+            'formats': formats,
+            'subtitles': subtitles,
```
```suggestion
            'subtitles': subtitles,
            **self.get_stream_metadata(),
```

you can inline self.get_stream_metadata() call

bashonly commented:
Please do not force push

bashonly commented:
Do not force push again

mp107 commented:
> Do not force push again

I've just read the message. Thank you for the information.

bashonly commented:
General preliminary review:

- Extractor methods should be private: `_get_stream_metadata()` not `get_stream_metadata()` etc
- Class vars are being overused. Avoid class constants that are only used once in the code. `channel_id` shouldn't be a class var
- Separating blocks of code with an empty line is acceptable, but here almost every other line is empty and it makes it harder to read.
- `ExtractorError` is being raised a lot. Shouldn't some of these be `expected=True`?
- Use relative imports: `from .common import InfoExtractor` etc
- Are these livestreams? Return `is_live` or `live_status` in the info dict if so.

mp107 commented:
Thank you for reviewing the code. I think I've fixed most of the issues (if not all of them).

>     * Extractor methods should be private: `_get_stream_metadata()` not `get_stream_metadata()` etc
Fixed.
 
>     * Class vars are being overused. Avoid class constants that are only used once in the code. `channel_id` shouldn't be a class var

I've left the access_token and channel_id as constants as they're being used in multiple methods: getting stream URL, stream metadata, logging in and extracting m3u8.

>     * Separating blocks of code with an empty line is acceptable, but here almost every other line is empty and it makes it harder to read.

I think I've fixed it. I've just left those around ifs and returns for code readability reasons.

>     * `ExtractorError` is being raised a lot. Shouldn't some of these be `expected=True`?

After further analyzing the API I've added handling login error (wrong username or password) with expected=True with the error message being passed to the user. Rest of them is related with more critical issues so I think I wouldn't be able to get rid of them.

>     * Use relative imports: `from .common import InfoExtractor` etc

I'm testing the plugin by sideloading the elemetal_tv.py file in `~/.config/yt-dlp/plugins/test/yt_dlp_plugins/extractor/` and making imports relative breaks it up. I don't know a workaround for it.

>     * Are these livestreams? Return `is_live` or `live_status` in the info dict if so.

I've fixed it.

I've also moved exception raises to method bodies related with loading the particular data the raise refers to to keep the main _real_extract method as clean as possible.

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+    }]
+
+    access_token = ''
+    channel_id = ''
+
+    def _get_channel_id(self, url):
+        url_parts = re.search('(?<=channel/)[0-9a-f]{24}', url)
+
+        if not url_parts or not url_parts.group(0):
+            return None
+
+        self.channel_id = url_parts.group(0)
+        self.write_debug(f'Channel ID: {self.channel_id}')
+
+        if not self.channel_id:
+            raise ExtractorError('Channel ID not found')
```
unnecessary
```suggestion
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+        },
+    }]
+
+    access_token = ''
```
```suggestion
    _access_token = None
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+
+    def _real_extract(self, url):
+        self._get_channel_id(url)
+        stream_url = self._get_stream_url()
```
the `_get_stream_url` code should just be inlined here

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+    def _real_extract(self, url):
+        self._get_channel_id(url)
+        stream_url = self._get_stream_url()
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(stream_url, self.channel_id, ext='mp4')
```
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            stream_url, channel_id, 'mp4')
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+            'is_live': True,
+            'formats': formats,
+            'subtitles': subtitles,
+            **self._get_stream_metadata(),
```
```suggestion
            **self._get_stream_metadata(channel_id),
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(stream_url, self.channel_id, ext='mp4')
+
+        return {
+            'id': self.channel_id,
```
```suggestion
            'id': channel_id,
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+
+
+class ElementalTVIE(InfoExtractor):
+    _LOGIN_REQUIRED = True
```
not used?
```suggestion
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+        if not self.channel_id:
+            raise ExtractorError('Channel ID not found')
+
+    def _get_stream_metadata(self):
```
```suggestion
    def _get_stream_metadata(self, channel_id):
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+                'Authorization': 'Bearer ' + self.access_token,
+            }
+            res_api = self._download_json(
+                'https://play.elemental.tv/v1/channels', self.channel_id, headers=headers)
```
```suggestion
                'https://play.elemental.tv/v1/channels', channel_id, headers=headers)
```

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+            }
+            res_api = self._download_json(
+                'https://play.elemental.tv/v1/channels', self.channel_id, headers=headers)
+            data = res_api.get('data').get(self.channel_id)
```
Don't use get chains. Use `traverse_obj`, or if it's meant to be fatal, just use brackets

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+            return {}
+
+    def _get_stream_url(self):
+        # Stream URL needs current epoch time rounded to 10000s
+        begin = int((time.time() - 60) / 10000) * 10000
```
Is this based on some of the site's javascript? `current_time` should look back two hours? A reference url would be nice to ensure correctness and for future maintenance

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+                'https://play.elemental.tv/v1/users/login', self.channel_id, data=json.dumps(post_data).encode()).get('data')
+        except ExtractorError as e:
+            if isinstance(e.cause, HTTPError) and e.cause.status == 400:
+                error_message = self._parse_json(e.cause.response.read().decode(), self.channel_id).get('error_info').get('description')
```
traverse_obj

bashonly commented on yt_dlp/extractor/elemental_tv.py:
```diff
@@ -0,0 +1,107 @@
+        except ExtractorError as e:
+            if isinstance(e.cause, HTTPError) and e.cause.status == 400:
+                error_message = self._parse_json(e.cause.response.read().decode(), self.channel_id).get('error_info').get('description')
+                raise ExtractorError(error_message, expected=True)
```
needs a raise for else
